pulser/simulation/simulation.py

```diff
@@ -15,116 +15,248 @@
         self._tot_duration = max(
             [self._seq._last(ch).tf for ch in self._seq._schedule]
         )
-
         if not (0 < sampling_rate <= 1.0):
             raise ValueError("The sampling rate (`sampling_rate` = "
                              f"{sampling_rate}) must be greater than 0 and "
                              "less than or equal to 1.")
-        if int(self._tot_duration*sampling_rate) < 4:
+        if int(self._tot_duration * sampling_rate) < 4:
             raise ValueError("`sampling_rate` is too small, less than 4 data "
                              "points.")
-        self.sampling_rate = sampling_rate
-
+        self._sampling_rate = sampling_rate
         self._qid_index = {qid: i for i, qid in enumerate(self._qdict)}
-        self.samples: dict[str, dict[str, dict]] = {
-            addr: {basis: {} for basis in ['ground-rydberg', 'digital']}
-            for addr in ['Global', 'Local']
-        }
-        self.operators = deepcopy(self.samples)
+        self._collapse_ops: list[qutip.Qobj] = []
+        self._times = self._adapt_to_sampling_rate(
+            np.arange(self._tot_duration, dtype=np.double)/1000)
+        self.evaluation_times = evaluation_times
+        self._bad_atoms: dict[Union[str, int], bool] = {
+            qid: False for qid in self._qid_index}
+        self._doppler_detune: dict[Union[str, int], float] = {
+            qid: 0. for qid in self._qid_index}
+        self.set_config(config) if config else self.set_config(SimConfig())
+        self.initial_state = 'all-ground'
+
+    @property
+    def config(self) -> SimConfig:
+        """Property getter for config."""
+        return self._config
+
+    def set_config(self, cfg: SimConfig) -> None:
+        """Sets current config to cfg and updates simulation parameters.
 
-        self._extract_samples()
-        self._build_basis_and_op_matrices()
+        Args:
+            cfg (SimConfig): New configuration.
+        """
+        if not isinstance(cfg, SimConfig):
+            raise ValueError(f"Object {cfg} is not a valid `SimConfig`")
+        self._config = cfg
+        if not ('SPAM' in self.config.noise and self.config.eta > 0):
+            self._bad_atoms = {qid: False for qid in self._qid_index}
+        if 'doppler' not in self.config.noise:
+            self._doppler_detune = {qid: 0. for qid in self._qid_index}
+        # Noise, samples and Hamiltonian update routine
         self._construct_hamiltonian()
+        if 'dephasing' in self.config.noise:
+            self._init_dephasing()
+
+    def update_config(self, config: SimConfig) -> None:
+        """Updates this SimConfig object with parameters of another one.
+
+        Mostly useful when dealing with multiple noise types in different
+        configurations and wanting to merge these configurations together.
+        Updates simulation parameters only when dealing with noise types that
+        weren't available in the former SimConfig.
```

Comment:
Well, if both simulations have `'SPAM'`as noise for example, the new `config`won't replace former `SPAM`parameters. I think we agreed on this ?

Reply:
Yes, but also if the new config has noise types that were not specified yet, it will add them. The phrasing you add here made it seem that you only changed parameters that were already defined in the old config.

Reply:
Might be an ambiguous use of the term "update". Perhaps distinguish between "adding" and "modifying" for clarity

Reply:
I agree then, your phrasing is clearer.